index.html

```diff
@@ -381,10 +432,13 @@
       swipe (direction) {
         if (!direction) throw new Error(`No/Invalid swipe direction provided: ${ direction }`);
         // capture current data
-        const prevState = this._state;
+        const prevState = this._data;
         // get new state
         const newState = this._calculateStateFromSwipe(direction)
-        this.state = newState; // TODO: move to updateState()?
+        this.data = newState; // TODO: move to an updateState()?
+
+        const hasLost = this._hasLost();
```

Comment:
is it expensive to call this AND _calculateStateFromSwipe each time? There's some overlap between what both of those methods do - theoretically, could we just have one function that each time will check the "swipeability" of the board each time, and return the lost state if all four are false?

Reply:
yea there is overlap btwn those 2 methods, but actually I think maybe the opportunity for processing savings comes from "combining" the fxnality of `_compareStates` and `_hasLost`. The fxnality of the `_merge{Direction}`s is quite specific and has a clear purpose (reconciling the row based on merge logic) as well as an output that would be tough to combine with logic like `_hasLost`. 

The merges are based on individual rows and in order to combine the 2, we would need to manage some state in the higher/calling method of `mergeRows` about the status of the other rows, which is the beginning of some messy logic.

However, considering `shouldIntroduce` & `hasLost` are already "self-contained" (no state management, just processing until the desired logic is proven out) & both looping over the board on each action and returning a value based on cell status (as opposed to returning a new board/row state) we could likely combine those 2 and simply return an object with variables attending to both concerns (if, after swiping, the user has lost and if there is a valid cell in which to add a new number).

Earlier in the game, we might lose some efficiency, because both of those methods would break very quickly (more likely to find `null` or a "mergeable pair") but we would likely gain efficiency as any individual game progresses.

The other thing to note is that, as it stands, while the complexity is `O(2*(n^2))` because of the twice-called, nested loops, the limit of how many cells need to be processed at any given time is upper-bound to 16, which should be fine for pretty much any CPU

Reply:
after trying to reconcile the 2 methods, we can get pretty close, but the issue seems to be in the order of operations. Here is the logic as I understand it:
- Swipe
- Check to see if a data point should be introduced (was there any change from previous board state)
- Introduce data point
- Check the latest state (including new cell data) to see if user has lost
- if lost, end game
- if !lost, continue

I don't think we can continue to refactor to introduce new cell data in the same method and then also evaluate if the player has lost without serious pretty messy logic. Consider the case where a data point is introduced in a cell that has already been evaluated - we would have to reevaluate that cell relative to its neighbors outside the evaluation loop, which is shitty. I think the separate evaluation methods (as is written in this PR), while more expensive, is the better approach.

Reply:
`_assess (prevState, currState) {
        const assessment = {
          shouldIntroduce: false, // if true after eval, add new number to data
          canProgress: false // if false after eval, user hasLost
        };
        // if there are no nulls && no mergeable neighbors, user has lost
        const Y_BOUND = currState.length;
        const X_BOUND = currState[0].length;
        
        for (let y = 0; y < Y_BOUND; y++) {
          for (let x = 0; x < X_BOUND; x++) {
            const currStateCell = currState[y][x];
            const prevStateCell = prevState[y][x];

            // NOTE: avoid out of bounds/reference errors on access
            //  - can't use `null` as that is the data value for 'empty cell'
            const u = y !== 0 ? currState[y-1][x] : undefined,
                  l = x !== 0 ? currState[y][x-1] : undefined,
                  d = y !== Y_BOUND-1 ? currState[y+1][x] : undefined,
                  r = x !== X_BOUND-1 ? currState[y][x+1] : undefined;
            
            if (currStateCell !== prevStateCell) assessment.shouldIntroduce = true;
            if (currStateCell === Cell.emptyValues.value || currStateCell === u || currStateCell === l || currStateCell === d || currStateCell === r) assessment.canProgress = true;
          }
        }

        return assessment;
      }`